File: java/src/main/java/com/a4bee/tddworkshop/service/BankingSystem.java
package com.a4bee.tddworkshop.service;

import java.util.HashMap;
import java.util.Map;

public class BankingSystem {
    private final Map<String, Double> accounts;

    public BankingSystem() {
        this.accounts = new HashMap<>();
    }

    public void createAccount(String accountNumber) {
        System.out.printf("Add account: %s%n", accountNumber);
        accounts.computeIfAbsent(accountNumber, (createAccountNumber -> 0.0d));
    }

    public void depositFunds(String accountNumber, double amount) {
        System.out.printf("Deposit founds: %.2f%n", amount);
        accounts.computeIfPresent(accountNumber,
                (depositAccountNumber, accountBalance) -> accountBalance + amount);
    }

    public void withdrawFunds(String accountNumber, double amount) {
        System.out.printf("Withdraw founds: %.2f%n", amount);
        if (accounts.containsKey(accountNumber) && accounts.get(accountNumber) >= amount) {
            accounts.put(accountNumber, accounts.get(accountNumber) - amount);
        }
    }

    public double checkBalance(String accountNumber) {
        double accountBalance = accounts.getOrDefault(accountNumber, 0.0);
        System.out.printf("Check balance: %.2f%n", accountBalance);
        return accountBalance;
    }

    public int getNumberOfAccounts() {
        return this.accounts.size();
    }
}

File: nodejs/service/BankingSystem.js
class BankingSystem {

  constructor() {
    this.accounts = new Map();
  }

  createAccount(accountNumber) {
    if (!this.accounts.has(accountNumber)) {
      console.log("Add account: %s", accountNumber);
      this.accounts.set(accountNumber, 0);
    }
  }

  depositFunds(accountNumber, amount) {
    if (this.accounts.has(accountNumber)) {
      console.log("Deposit founds: %d", amount)
      this.accounts.set(accountNumber, this.accounts.get(accountNumber) + amount);
    }
  }

  withdrawFunds(accountNumber, amount) {
    if (this.accounts.has(accountNumber) && this.accounts.get(accountNumber) >= amount) {
      console.log("Withdraw founds: %d", amount)
      this.accounts.set(accountNumber, this.accounts.get(accountNumber) - amount);
    }
  }

  checkBalance(accountNumber) {
    const accountBalance = this.accounts.get(accountNumber) || 0;
    console.log("Account balance: %d", accountBalance);
    return accountBalance;
  }

  numberOfAccounts() {
    return this.accounts.size;
  }
}

export default BankingSystem;
